src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ConcurrencyLimiter.cs

```diff
@@ -0,0 +1,300 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+
+namespace System.Threading.RateLimiting
+{
+    /// <summary>
+    /// <see cref="RateLimiter"/> implementation that helps manage concurrent access to a resource.
+    /// </summary>
+    public sealed class ConcurrencyLimiter : RateLimiter
+    {
+        private int _permitCount;
+        private int _queueCount;
+        private bool _disposed;
+
+        private readonly ConcurrencyLimiterOptions _options;
+        private readonly Deque<RequestRegistration> _queue = new Deque<RequestRegistration>();
+
+        private static readonly ConcurrencyLease SuccessfulLease = new ConcurrencyLease(true, null, 0);
+        private static readonly ConcurrencyLease FailedLease = new ConcurrencyLease(false, null, 0);
+        private static readonly ConcurrencyLease QueueLimitLease = new ConcurrencyLease(false, null, 0, "Queue limit reached");
+
+        // Use the queue as the lock field so we don't need to allocate another object for a lock and have another field in the object
+        private object Lock => _queue;
+
+        /// <summary>
+        /// Initializes the <see cref="ConcurrencyLimiter"/>.
+        /// </summary>
+        /// <param name="options">Options to specify the behavior of the <see cref="ConcurrencyLimiter"/>.</param>
+        public ConcurrencyLimiter(ConcurrencyLimiterOptions options)
+        {
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _permitCount = _options.PermitLimit;
+        }
+
+        /// <inheritdoc/>
+        public override int GetAvailablePermits() => _permitCount;
+
+        /// <inheritdoc/>
+        protected override RateLimitLease AcquireCore(int permitCount)
+        {
+            // These amounts of resources can never be acquired
+            if (permitCount > _options.PermitLimit)
+            {
+                throw new ArgumentOutOfRangeException(nameof(permitCount), permitCount, SR.Format(SR.PermitLimitExceeded, permitCount, _options.PermitLimit));
+            }
+
+            // Return SuccessfulLease or FailedLease to indicate limiter state
+            if (permitCount == 0 && !_disposed)
```

Comment:
Do we want to throw ObjectDisposedException if this has been disposed?

Reply:
Should an ODE throw synchronously or be wrapped in a ValueTask?

Reply:
I guess the first question is:

* Do we want to throw ODE or always return "failed" when a limiter is disposed?

Doing a quick check of other System.Threading disposable types, they seem to throw ODE: Barrier, ManualResetEventSlim, ReaderWriterLockSlim


For the next question, I would say it is similar to argument checking - so we would throw inline.